File: 0081-search-in-rotated-sorted-array-ii/0081-search-in-rotated-sorted-array-ii.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {boolean}
 */
var search = function(nums, target) {

    let low = 0;
    const n = nums.length;
    let high = n - 1;
    while (low <= high) {
        let mid = Math.floor((low + high) / 2);
        
        if (nums[mid] == target) return true;
        else if(nums[mid]==nums[low] && nums[mid]==nums[high]){
            low++;
            high--;
            continue;
        }
        else if (nums[mid] >= nums[low]) {
            // if left part is sorted
            if (target >= nums[low] && target <= nums[mid]) {
                high = mid - 1;

            }
            else {
                low = mid + 1;
            }
        }
        else {
            // if right is sorted
            if (target >= nums[mid] && target <= nums[high]) {

                low = mid + 1;

            }
            else {
                high = mid - 1;

            }
        }
    }
    return false;

    
};

File: 0081-search-in-rotated-sorted-array-ii/0081-search-in-rotated-sorted-array-ii.cpp
class Solution {
public:
    bool search(vector<int>& arr, int val) {
        int n = arr.size();
        int low = 0;
        int high = n - 1;
        while (low <= high) {
            int mid = (low + high) / 2;
            if (arr[mid] == val) {
                return true;
            } else if (arr[mid] == arr[low] && arr[low] == arr[high]) {
                low++;
                high--;
            }
            // left part sorted hai
            else if (arr[mid] >= arr[low]) {
                if (val >= arr[low] && val <= arr[mid]) {
                    high = mid - 1;
                } else {
                    low = mid + 1;
                }
            }
            // toh matlab right part sorted hoga
            else {
                if (val >= arr[mid] && val <= arr[high]) {
                    low = mid + 1;
                } else {
                    high = mid - 1;
                }
            }
        }
        return false;
    }
};